src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,286 @@
+"""Construct geometric graphs from data for multiscale clustering."""
+
+import numpy as np
+
+from scipy.spatial.distance import pdist, squareform
+from scipy.sparse.csgraph import minimum_spanning_tree
+from scipy.sparse import csr_matrix
+
+from pygenstability.pygenstability import run as pgs_run
+from pygenstability.plotting import plot_scan as pgs_plot_scan
+from pygenstability.optimal_scales import identify_optimal_scales
+
+
+def compute_kNN(D, k=5):
+    """Computes kNN graph."""
+
+    N = D.shape[0]
+
+    # get k nearest neighbours for each point
+    k_neighbours = np.argsort(D, axis=1)[:, 1 : k + 1]
+
+    # initialise adjacency matrix
+    A = np.zeros((N, N))
+
+    # build kNN graph
+    for i in range(N):
+        for neighbour in k_neighbours[i]:
+            A[i, neighbour] = D[i, neighbour]
+            A[neighbour, i] = A[i, neighbour]
+
+    return A
+
+
+def compute_CkNN(D, k=5, delta=1):
```

Comment:
is this the same as in https://github.com/chlorochrule/cknn ? it looks a bit more complicated there, no? I'm not so familiar with the details of this code, just asking!

Reply:
Yes, it's the same. But implementing ourselves has the advantage that users don't need to install this additional package.

Reply:
Ok, very good!